### ISSUE TITLE
Model generates audio with random content during inference

### ISSUE THREAD

**Issue by huutuongtu** [SWivid/F5-TTS#246]
Hi, thanks for the great work. I trained the model on my own dataset (Vietnamese, 1 speaker, 1.5 hours, just for testing) for about 70k steps. In [this comment,](https://github.com/SWivid/F5-TTS/issues/9#issuecomment-2405307421) you mentioned that at a loss of around 0.6, you should start hearing something intelligible. My current loss is around ~0.3 to 0.4. However, when I run inference, the generated audio seems to contain random content, although the speaker similarity is good. I also rechecked the text input and tokenizer, and they seem to be working well.
What could be the problem?
![W B Chart 10_24_2024, 3_45_33 PM](https://github.com/user-attachments/assets/cdaa1135-5712-436c-9601-685facc227df)

**Comment by SWivid** [comment]
So you are training from scratch or finetuning on pretrained model.
A probably reason is: the low loss is for overfitting if you use small batchsize.

**Comment by huutuongtu** [comment]
> So you are training from scratch or finetuning on pretrained model. A possible reason is: the low loss is for overfitting if you use small batchsize.

I trained a model from scratch using two A100 40GB GPUs. Below is my configuration:

```
learning_rate = 7.5e-5
batch_size_per_gpu = 9600 #can't be larger as it will cause OOM
batch_size_type = "frame"
max_samples = 32
grad_accumulation_steps = 1
max_grad_norm = 1.
epochs = 3000
num_warmup_updates = 2000
save_per_updates = 1000
last_per_steps = 2000
```

I also tried debugging using this code in the cfm.py forward function to ensure correct tokenization.

```
# handle text as string
if isinstance(text, list):
    if exists(self.vocab_char_map):
        print(text[0])
        print(text[1])
        text = list_str_to_idx(text, self.vocab_char_map).to(device) # indices
        print(text[0])
        print(text[1])
    else:
        text = list_str_to_tensor(text).to(device)
    assert text.shape[0] == batch
```
Return: 
```
['c', 'h', 'o', ' ', 'n', 'ó', ' ', 'c', 'á', 'i', ' ', 'q', 'u', 'á', 'i', ' ', 'x', 'ì', ' ', 't', 'h', 'ế']
['t', 'ô', 'i', ' ', 'v', 'ừ', 'a', ' ', 'n', 'h', 'ậ', 'n', ' ', 'đ', 'ư', 'ợ', 'c', ' ', 'k', 'ế', 't', ' ', 'q', 'u', 'ả']
tensor([ 5, 10, 17,  0, 16, 37,  0,  5, 28, 11,  0, 19, 23, 28, 11,  0, 25, 34,
         0, 22, 10, 64, -1, -1, -1, -1, -1, -1, -1], device='cuda:0')
tensor([22, 38, 11,  0, 24, 86,  3,  0, 16, 10, 55, 16,  0, 44, 48, 82,  5,  0,
        13, 64, 22,  0, 19, 23, 50, -1, -1, -1, -1], device='cuda:0')
```

I also tried inferring using ref_audio and ref_text from the training set, and I set gen_text to ref_text to check for overfitting, but the generated audio still seems random.

**Comment by SWivid** [comment]
@huutuongtu I would recommend a smaller model size.
And just train longer, at least 200k updates thought to hear something reasonable, cuz we have no phoneme-level force-alignment 
(if you're interested in the mechanism, feel free to see our arxiv paper, also probably E2 TTS and voicebox), basically it's just harder to train without it

**Comment by huutuongtu** [comment]
> @huutuongtu I would recommend a smaller model size. And just train longer, at least 200k updates thought to hear something reasonable, cuz we have no phoneme-level force-alignment (if you're interested in the mechanism, feel free to see our arxiv paper, also probably E2 TTS and voicebox), basically it's just harder to train without it

Hmm... It seems that training longer with the small old dataset still couldn't generate audio with clear content. I tried training with a larger dataset (about 12 hours), and the generated audio now seems to have better quality. So, you were right - the problem is that this model struggles to learn alignment between text and audio.

**Comment by MattLiutt** [comment]
@huutuongtu may I know how you prepare your dataset, as I'm trying to finetune with gradio app with opensource like google/fleurs, but it seems cannot do transcribe step. It would be helpful if you can provide some insights. Thanks!

**Comment by huutuongtu** [comment]
> @huutuongtu may I know how you prepare your dataset, as I'm trying to finetune with gradio app with opensource like google/fleurs, but it seems cannot do transcribe step. It would be helpful if you can provide some insights. Thanks!

I used an older version of this repository, which has since undergone many changes. To train the model, here are some basics you need. You will need a dataset with audio and transcript annotations. You will also need to extend the vocabulary. Additionally, you should re-check the tokenization function to ensure it correctly tokenizes the text into indices (e.g., character or word tokenization). For example, when preparing a dataset that is not in Mandarin, this line may cause issues with your model: [convert_char_to_pinyin](https://github.com/SWivid/F5-TTS/blob/be53fb179c17c8db16656d017af26578633cd343/src/f5_tts/train/datasets/prepare_csv_wavs.py#L47)

All training and fine-tuning details are listed here: https://github.com/SWivid/F5-TTS/tree/main/src/f5_tts/train

**Comment by MattLiutt** [comment]
thanks for the reply! So basically you didn't create a customized vocab/token for Vietnamese but extend the vocab. May I know how to check the tokenization function?

**Comment by huutuongtu** [comment]
> thanks for the reply! So basically you didn't create a customized vocab/token for Vietnamese but extend the vocab. May I know how to check the tokenization function?

1. I created my own tokenizer for Vietnamese and trained it with a small dataset. It generated random content, such as this title. I tried training with a larger dataset, and this issue no longer exists. You can check the vocab format [here](https://github.com/SWivid/F5-TTS/blob/main/data/Emilia_ZH_EN_pinyin/vocab.txt). 
2. But, new issues have appeared, such as mispronunciations. My dataset may not be large enough (Not sure). 
3. The tokenizer function is _list_str_to_idx_ or _list_str_to_tensor_ in [utils.py](https://github.com/SWivid/F5-TTS/blob/be53fb179c17c8db16656d017af26578633cd343/src/f5_tts/model/utils.py#L88) (Depending on which tokenizer you use, you can check [here](https://github.com/SWivid/F5-TTS/blob/be53fb179c17c8db16656d017af26578633cd343/src/f5_tts/model/cfm.py#L116).

**Comment by MattLiutt** [comment]
> > thanks for the reply! So basically you didn't create a customized vocab/token for Vietnamese but extend the vocab. May I know how to check the tokenization function?
> 
> 1. I created my own tokenizer for Vietnamese and trained it with a small dataset. It generated random content, such as this title. I tried training with a larger dataset, and this issue no longer exists. You can check the vocab format [here](https://github.com/SWivid/F5-TTS/blob/main/data/Emilia_ZH_EN_pinyin/vocab.txt).
> 2. But, new issues have appeared, such as mispronunciations. My dataset may not be large enough (Not sure).
> 3. The tokenizer function is _list_str_to_idx_ or _list_str_to_tensor_ in [utils.py](https://github.com/SWivid/F5-TTS/blob/be53fb179c17c8db16656d017af26578633cd343/src/f5_tts/model/utils.py#L88) (Depending on which tokenizer you use, you can check [here](https://github.com/SWivid/F5-TTS/blob/be53fb179c17c8db16656d017af26578633cd343/src/f5_tts/model/cfm.py#L116).

Appreciated! Definitely helpful to my training!

**Comment by MattLiutt** [comment]
Just one more question, you mentioned you create tokenizer yourself. Does it mean that you are not using their vocab.txt and use as pinyin like [function here](https://github.com/SWivid/F5-TTS/blob/be53fb179c17c8db16656d017af26578633cd343/src/f5_tts/model/utils.py#L137). Thanks!

**Comment by huutuongtu** [comment]
> Just one more question, you mentioned you create tokenizer yourself. Does it mean that you are not using their vocab.txt and use as pinyin like [function here](https://github.com/SWivid/F5-TTS/blob/be53fb179c17c8db16656d017af26578633cd343/src/f5_tts/model/utils.py#L137). Thanks!

Yes. New vocab and not use pinyin convert.

**Comment by MattLiutt** [comment]
> Yes. New vocab and not use pinyin convert.

Thanks! You said that you make use of tokenizer function is _list_str_to_idx_ or _list_str_to_tensor_ in [utils.py](https://github.com/SWivid/F5-TTS/blob/be53fb179c17c8db16656d017af26578633cd343/src/f5_tts/model/utils.py#L88). May I know how you make use of the utils. Great appreciated!

**Comment by huutuongtu** [comment]
> > Yes. New vocab and not use pinyin convert.
> 
> Thanks! You said that you make use of tokenizer function is _list_str_to_idx_ or _list_str_to_tensor_ in [utils.py](https://github.com/SWivid/F5-TTS/blob/be53fb179c17c8db16656d017af26578633cd343/src/f5_tts/model/utils.py#L88). May I know how you make use of the utils. Great appreciated!

In [train.py](https://github.com/SWivid/F5-TTS/blob/be53fb179c17c8db16656d017af26578633cd343/src/f5_tts/train/train.py#L58) the author pass vocab_char_map to model so 
`text = list_str_to_idx(text, self.vocab_char_map).to(device)` is used
